README.md
# boilerplate-nodejs

Node.js プロジェクト用のボイラープレート

## GitHub

[![nodejs ci](https://github.com/kannkyo/boilerplate-nodejs/actions/workflows/nodejs-ci.yml/badge.svg)](https://github.com/kannkyo/boilerplate-nodejs/actions/workflows/nodejs-ci.yml)
[![nodejs publish](https://github.com/kannkyo/boilerplate-nodejs/actions/workflows/nodejs-publish.yml/badge.svg)](https://github.com/kannkyo/boilerplate-nodejs/actions/workflows/nodejs-publish.yml)
[![CodeQL](https://github.com/kannkyo/boilerplate-nodejs/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/kannkyo/boilerplate-nodejs/actions/workflows/codeql-analysis.yml)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.4](https://github.com/kannkyo/boilerplate-nodejs/compare/v0.0.3...v0.0.4) (2022-07-10)

### [0.0.3](https://github.com/kannkyo/boilerplate-nodejs/compare/v0.0.2...v0.0.3) (2022-07-10)


### Added

* add nodejs-cron.yml ([7770f3e](https://github.com/kannkyo/boilerplate-nodejs/commit/7770f3e7fb89bca4319477467c02b019a65d6cd0))


### Changed

* rename sbom file name ([bc116bc](https://github.com/kannkyo/boilerplate-nodejs/commit/bc116bc9ab084c83cee24b4d2173306438819dc6))


### Fixed

* npm audit found vulnerabilities [#10](https://github.com/kannkyo/boilerplate-nodejs/issues/10) ([a7610e5](https://github.com/kannkyo/boilerplate-nodejs/commit/a7610e5301f1bb83632879fb8390bfe7df6be270))
* package.json & package-lock.json to reduce vulnerabilities ([5634569](https://github.com/kannkyo/boilerplate-nodejs/commit/56345695826dd810412133cf17b84f87a586fd24))
* upgrade @cyclonedx/bom from 3.3.1 to 3.10.0 ([3fb192a](https://github.com/kannkyo/boilerplate-nodejs/commit/3fb192a95ad16465fa13bd42b9c9b7f71ad428df))
* upgrade cookie-parser from 1.4.5 to 1.4.6 ([32d5cb5](https://github.com/kannkyo/boilerplate-nodejs/commit/32d5cb5f4d141d03563b59e81f017d5d3ee31ce7))
* upgrade debug from 4.3.2 to 4.3.3 ([c324c0e](https://github.com/kannkyo/boilerplate-nodejs/commit/c324c0e4fef3b9a09d4634863c99c9b7e856e98e))
* upgrade debug from 4.3.3 to 4.3.4 ([6529b86](https://github.com/kannkyo/boilerplate-nodejs/commit/6529b8626fb06e7d4eceb5adf144fdb715e8db33))
* upgrade express from 4.17.2 to 4.17.3 ([ea8ab42](https://github.com/kannkyo/boilerplate-nodejs/commit/ea8ab42dc0c5df8fcc5808589412ff221934f366))
* upgrade express from 4.17.3 to 4.18.1 ([0a3f5d2](https://github.com/kannkyo/boilerplate-nodejs/commit/0a3f5d26fb0605a04e2498e790a7740cd37c3591))
* upgrade http-errors from 1.8.0 to 1.8.1 ([0a38ebc](https://github.com/kannkyo/boilerplate-nodejs/commit/0a38ebcb5d0d65cf3f02a090a5ca16397493fd67))
* upgrade standard-version from 9.3.2 to 9.5.0 ([bb48f39](https://github.com/kannkyo/boilerplate-nodejs/commit/bb48f39adedb93186a8bb73317568f92e7adab5d))

### [0.0.2](https://github.com/kannkyo/boilerplate-nodejs/compare/v0.0.1...v0.0.2) (2021-11-06)


### Changed

* **github:** refactor github actions ([73a52dd](https://github.com/kannkyo/boilerplate-nodejs/commit/73a52ddcd6f2db28d8a3a67c651ccfe4485a1003))

### [0.0.1](https://github.com/kannkyo/boilerplate-nodejs/compare/v0.0.0...v0.0.1) (2021-11-06)


### Added

* @cyclonedx/bom ([794d42e](https://github.com/kannkyo/boilerplate-nodejs/commit/794d42eb83f10dbb0d5756c9c57443c463b4d6e7))
* **docs:** add sample docs ([9c13159](https://github.com/kannkyo/boilerplate-nodejs/commit/9c13159337dadc7efa24fc137c3934fe82e4ba2f))

## 0.0.0 (2021-08-25)


### Added

* add .npmignore ([dfea58e](https://github.com/kannkyo/boilerplate-nodejs/commit/dfea58e8fdd1bb2131c407b1849df859363daa45))
* add .versionrc ([ee88a43](https://github.com/kannkyo/boilerplate-nodejs/commit/ee88a4343bc7edc6bd1814236a39e08ca1c06e86))


test/app.spec.js
const request = require('supertest')
const sinon = require('sinon')
const rewire = require('rewire')
const assert = require('assert')
const app = rewire('../app')

describe('app test', () => {
  it('GET invalid path', (done) => {
    request(app)
      .get('/hogehoeg')
      .expect(404)
      .end(done)
  })

  it('Call error handler with 500 error', (done) => {
    const err = {
      message: 'test message',
      status: 404
    }
    const req = {
      app: {
        get: sinon.spy()
      }
    }
    const res = {
      locals: {
        message: '',
        error: ''
      },
      status: sinon.spy(),
      render: sinon.spy()
    }
    const next = sinon.spy()

    app.__get__('errorHandler')(err, req, res, next)

    assert.strictEqual(res.locals.message, 'test message')
    assert.notStrictEqual(res.locals.error, {})
    assert.ok(res.render.withArgs('error').calledOnce)
    assert.ok(res.status.withArgs(404).calledOnce)

    done()
  })

  it('Call error handler with 404 error', (done) => {
    const err = {
      message: 'test message',
      status: null
    }
    const req = {
      app: {
        get: sinon.spy()
      }
    }
    const res = {
      locals: {
        message: '',
        error: ''
      },
      status: sinon.spy(),
      render: sinon.spy()
    }
    const next = sinon.spy()

    app.__get__('errorHandler')(err, req, res, next)

    assert.strictEqual(res.locals.message, 'test message')
    assert.notStrictEqual(res.locals.error, {})
    assert.ok(res.render.withArgs('error').calledOnce)
    assert.ok(res.status.withArgs(500).calledOnce)

    done()
  })
})


test/index.spec.js
const path = require('path')
const express = require('express')
const request = require('supertest')
const sinon = require('sinon')
const rewire = require('rewire')
const assert = require('assert')
const index = rewire('../routes/index')

const app = express()
app.set('views', path.join(__dirname, '../', 'views'))
app.set('view engine', 'jade')
app.use(express.static(path.join(__dirname, '../', 'public')))
app.use('/', index)

describe('index test', () => {
  it('GET /', (done) => {
    request(app)
      .get('/')
      .expect(200, '<!DOCTYPE html><html><head><title>Express</title><link rel="stylesheet" href="/stylesheets/style.css"></head><body><h1>Express</h1><p>Welcome to Express</p></body></html>')
      .end(done)
  })

  it('Call index without exception', (done) => {
    const req = {}
    const res = {
      render: sinon.spy()
    }
    const next = sinon.spy()

    index.__get__('index')(req, res, next)

    assert.ok(res.render.calledOnce)

    done()
  })

  it('Call index with exception', (done) => {
    // https://codewithhugo.com/express-request-response-mocking/
    // https://www.it-swarm-ja.tech/ja/javascript/mocha%E3%81%8A%E3%82%88%E3%81%B3nodejs%E3%82%92%E4%BD%BF%E7%94%A8%E3%81%97%E3%81%9F%E3%83%97%E3%83%A9%E3%82%A4%E3%83%99%E3%83%BC%E3%83%88%E9%96%A2%E6%95%B0%E3%81%AE%E5%8D%98%E4%BD%93%E3%83%86%E3%82%B9%E3%83%88/1045289963/
    const req = {}
    const res = {
      render: sinon.fake.throws(new Error('test message'))
    }
    const next = sinon.spy()

    index.__get__('index')(req, res, next)

    assert.ok(next.called)

    done()
  })
})


test/users.spec.js
const sinon = require('sinon')
const rewire = require('rewire')
const path = require('path')
const express = require('express')
const request = require('supertest')
const assert = require('assert')
const users = rewire('../routes/users')

const app = express()
app.set('views', path.join(__dirname, '../', 'views'))
app.set('view engine', 'jade')
app.use(express.static(path.join(__dirname, '../', 'public')))
app.use('/users', users)

describe('users test', () => {
  it('GET /users', (done) => {
    request(app)
      .get('/users')
      .expect(200, 'respond with a resource')
      .end(done)
  })

  it('Call users without exception', (done) => {
    const req = {}
    const res = {
      send: sinon.spy()
    }
    const next = sinon.spy()

    users.__get__('users')(req, res, next)

    assert.ok(res.send.calledOnce)

    done()
  })

  it('Call users with exception', (done) => {
    const req = {}
    const res = {
      send: sinon.fake.throws(new Error('test message'))
    }
    const next = sinon.spy()

    users.__get__('users')(req, res, next)

    assert.ok(next.called)

    done()
  })
})


app.js
/**
 * main関数
 * @module app
 */

const createError = require('http-errors')
const express = require('express')
const path = require('path')
const cookieParser = require('cookie-parser')
const logger = require('morgan')

const indexRouter = require('./routes/index')
const usersRouter = require('./routes/users')

const app = express()

// view engine setup
app.set('views', path.join(__dirname, 'views'))
app.set('view engine', 'jade')

app.use(logger('dev'))
app.use(express.json())
app.use(express.urlencoded({ extended: false }))
app.use(cookieParser())
app.use(express.static(path.join(__dirname, 'public')))

app.use('/', indexRouter)
app.use('/users', usersRouter)

// catch 404 and forward to error handler
app.use((req, res, next) => {
  next(createError(404))
})

// error handler
const errorHandler = (err, req, res, next) => {
  // set locals, only providing error in development
  res.locals.message = err.message
  res.locals.error = req.app.get('env') === 'development' ? err : {}

  // render the error page
  res.status(err.status || 500)
  res.render('error')
}

app.use(errorHandler)

module.exports = app
